### ISSUE TITLE
Possible to add check for nix.conf validity before switching configuration?

### ISSUE THREAD

**Issue by nh2** [NixOS/nixpkgs#29592]
## Issue description

I use `nixops` and accidentally deployed

```nix
{ nix.extraOptions "ssh-substituter-hosts myhost"; }
```

(missing an `=`) instead of

```nix
{ nix.extraOptions = "ssh-substituter-hosts = myhost"; }
```

Now my `nix.conf` contains invalid contents, and `nix-copy-closure` fails with

```
error: illegal configuration line ‘ssh-substituter-hosts myhost’ in ‘/etc/nix/nix.conf’
```

so I cannot use nixops any more to fix the problem, and have to fix each host individually.

Is it possible to somehow add a check phase so that switching to a configuratin will fail if the `nix` of that configuration can't understand the `nix.conf` of that configuration, to prevent this issue?

**Comment by copumpkin** [comment]
@edolstra now that Nix actually knows valid option names in a first-class manner, perhaps a `nix check-config` subcommand in the 1.12 new UI would be a good addition? Then we could feed generated config files into Nix and see if they fail which would cause the entire switch to fail.

Alternately (or additionally), we could just make `extraOptions` into an attrset to minimize potential for error, since there's not really any case I can think of where you'd want freeform text.

**Comment by dtzWill** [comment]
Just ran into this, and not for the first time... let's see if it can be the last time I do a dance kludging around a broken /etc/nix/nix.conf (I always forget you can chase down a working generation and switch/activate it).

> nix check-config subcommand in the 1.12 new UI

This would be great.

> Alternately (or additionally), we could just make extraOptions into an attrset to minimize potential for error, since there's not really any case I can think of where you'd want freeform text.

Agreed! Although we may want to instead introduce a new option (not sure what to call it)--so that the existing `extraOptions` usage doesn't break and "in case" attributes are somehow deficient in the future or in ways we didn't think of.

**Comment by matthewbauer** [comment]
I think this was resolved recently?

**Comment by nh2** [comment]
Where?

**Comment by matthewbauer** [comment]
in this commit: https://github.com/NixOS/nixpkgs/commit/fa1d1eec6e5ef9f248416bec9df10ee2c9ed0e15

**Comment by matthewbauer** [comment]
fixed in fa1d1eec6e5ef9f248416bec9df10ee2c9ed0e15
